src/common/pkcs11_lib.c

```diff
@@ -1824,6 +1825,11 @@ int sign_value(pkcs11_handle_t *h, cert_object_t *cert, CK_BYTE *data,
       /* increase signature length as long as it it to short */
       free(*signature);
       *signature = NULL;
+      if (current_signature_length >= *signature_length) {
+        /* workaround for buggy PKCS#11 implementation: it didn't change
+           or even lowered buffer size - forcing using larger (double size) buffer */
+        *signature_length = current_signature_length * 2;
+      }
```

Comment:
Well, theoretically we could have here endless loop, but only in the case when this buggy PKCS#11 library returns always `CKR_BUFFER_TOO_SMALL` even for large buffer sizes. I mean PKCS#11 library did check size and returned this error, so I think we can assume it was done intentionally. Bug here is not updating `signature_length` parameter, not intentional messing with returned data.

Yes, we can check maximum value here, but what size we can select? 65536 bytes?

Reply:
Yes, see https://github.com/OpenSC/pam_pkcs11/commit/7ce58a6c6cdd8f6b90fffe552d2b7137677da21a .